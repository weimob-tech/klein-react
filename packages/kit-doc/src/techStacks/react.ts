// import type { IDumiTechStack } from '@/types';
import {
  transformSync,
  type ExportDefaultDeclaration,
  type ExportDefaultExpression,
  type Expression,
  type ImportDeclaration,
  type ObjectPattern,
  type ReturnStatement,
  type Span,
  type VariableDeclaration,
} from '@swc/core';
import Visitor from '@swc/core/Visitor';

type IDumiTechStack = any;

function createReturnStmt(exp: Expression, span: Span): ReturnStatement {
  return {
    type: 'ReturnStatement',
    span,
    argument: {
      type: 'ObjectExpression',
      span,
      properties: [
        {
          type: 'KeyValueProperty',
          key: {
            type: 'Identifier',
            span,
            value: 'default',
            optional: false,
          },
          value: exp,
        },
      ],
    },
  };
}

/**
 * swc plugin for replace export to return, and replace import to await import
 */
class ReactDemoVisitor extends Visitor {
  // @ts-ignore
  visitImportDeclaration(
    n: ImportDeclaration,
  ): ImportDeclaration | VariableDeclaration {
    if (!n.typeOnly) {
      const namespaceImport = n.specifiers.find(
        (s) => s.type === 'ImportNamespaceSpecifier',
      );
      const id = namespaceImport
        ? namespaceImport.local
        : ({
            type: 'ObjectPattern',
            span: n.span,
            properties: n.specifiers.map((s) => {
              if (
                s.type === 'ImportDefaultSpecifier' ||
                (s.type === 'ImportSpecifier' &&
                  s.imported?.type === 'Identifier')
              ) {
                return {
                  type: 'KeyValuePatternProperty',
                  span: s.span,
                  key:
                    s.type === 'ImportSpecifier'
                      ? s.imported!
                      : {
                          type: 'Identifier',
                          span: s.span,
                          value: 'default',
                          optional: false,
                        },
                  value: s.local,
                };
              }

              return {
                type: 'AssignmentPatternProperty',
                span: s.span,
                key: s.local,
              };
            }),
            optional: false,
          } as ObjectPattern);

      return {
        type: 'VariableDeclaration',
        kind: 'const',
        declare: false,
        span: n.span,
        declarations: [
          {
            type: 'VariableDeclarator',
            span: n.span,
            definite: false,
            id,
            init: {
              span: n.span,
              type: 'AwaitExpression',
              argument: {
                type: 'CallExpression',
                span: n.span,
                callee: {
                  type: 'Import',
                  span: n.span,
                },
                arguments: [{ expression: n.source }],
              },
            },
          },
        ],
      };
    }

    return n;
  }

  // @ts-ignore
  visitExportDefaultDeclaration(
    n: ExportDefaultDeclaration,
  ): ExportDefaultDeclaration | ReturnStatement {
    if (n.decl.type !== 'TsInterfaceDeclaration') {
      return createReturnStmt(n.decl, n.span);
    }

    return n;
  }

  // @ts-ignore
  visitExportDefaultExpression(n: ExportDefaultExpression): ReturnStatement {
    return createReturnStmt(n.expression, n.span);
  }

  visitTsType(n: any) {
    // to avoid `Error: Method visitTsType not implemented.`
    // ref: https://github.com/swc-project/swc/blob/31630ba913fffa68fc4703cf6497d09f2e6ba2e9/node-swc/src/Visitor.ts#L1680
    return n;
  }
}

export default class ReactTechStack implements IDumiTechStack {
  name = 'react';

  isSupported(...[, lang]: Parameters<IDumiTechStack['isSupported']>) {
    return ['jsx', 'tsx'].includes(lang);
  }

  transformCode(...[raw, opts]: Parameters<IDumiTechStack['transformCode']>) {
    if (opts.type === 'code-block') {
      const isTSX = opts.fileAbsPath.endsWith('.tsx');
      const { code } = transformSync(raw, {
        filename: opts.fileAbsPath,
        jsc: {
          parser: {
            syntax: isTSX ? 'typescript' : 'ecmascript',
            [isTSX ? 'tsx' : 'jsx']: true,
          },
          target: 'es2022',
        },
        module: {
          type: 'es6',
        },
        plugin: (m) => new ReactDemoVisitor().visitProgram(m),
      });

      return `React.lazy(async () => {
${code}
})`;
    }

    return raw;
  }
}
