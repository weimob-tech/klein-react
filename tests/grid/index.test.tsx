import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Grid } from 'klein';

const { Col, Row } = Grid;

describe('Gird', () => {
  test('should render col', () => {
    const { asFragment } = render(<Col />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render row', () => {
    const { asFragment } = render(<Row />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render wrapped row correctly', () => {
    const { asFragment } = render(
      <Row>
        <div>
          <Col />
        </div>
      </Row>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('when typeof gutter is number', () => {
    render(<Row gutter={8} />);
    expect(document.querySelector('.klein-row')).toHaveStyle(
      'margin-left: -4px; margin-right: -4px;',
    );
    const { getByText } = render(
      <Row gutter={8}>
        <Col>col-8</Col>
      </Row>,
    );
    expect(getByText('col-8')).toHaveStyle(
      'padding-left: 4px; padding-right: 4px;',
    );
  });

  test('when typeof gutter is object', () => {
    render(<Row gutter={{ xs: 8, sm: 16 }}>gutter</Row>);
    expect(document.querySelector('.klein-row')).toHaveStyle(
      'margin-left: -4px; margin-right: -4px;',
    );
    const { getByText } = render(
      <Row gutter={8}>
        <Col>col-8</Col>
      </Row>,
    );
    expect(getByText('col-8')).toHaveStyle(
      'padding-left: 4px; padding-right: 4px;',
    );
  });

  it('when typeof gutter is object array', () => {
    render(
      <Row
        gutter={[
          { xs: 8, sm: 16, md: 24, lg: 32, xl: 40 },
          { xs: 8, sm: 16, md: 24, lg: 32, xl: 40 },
        ]}
      />,
    );
    expect(document.querySelector('.klein-row')).toHaveStyle(
      'margin-left: -4px; margin-right: -4px;',
    );
  });

  it('when typeof gutter is array', () => {
    render(<Row gutter={[8, 16]} />);
    expect(document.querySelector('.klein-row')).toHaveStyle(
      'margin-left: -4px; margin-right: -4px;',
    );
  });
});
