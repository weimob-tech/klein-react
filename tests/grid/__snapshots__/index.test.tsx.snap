// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gird should render col 1`] = `
<DocumentFragment>
  <div
    class="klein-col"
  />
</DocumentFragment>
`;

exports[`Gird should render row 1`] = `
<DocumentFragment>
  <div
    class="klein-row klein-row-start"
  />
</DocumentFragment>
`;

exports[`Gird should render wrapped row correctly 1`] = `
<DocumentFragment>
  <div
    class="klein-row klein-row-start"
  >
    <div>
      <div
        class="klein-col"
        style="padding-left: 0px; padding-right: 0px;"
      />
    </div>
  </div>
</DocumentFragment>
`;
